python/beeai_framework/agents/base.py

```diff
@@ -1,28 +1,85 @@
 # Copyright 2025 © BeeAI a Series of LF Projects, LLC
 # SPDX-License-Identifier: Apache-2.0
 
-from abc import ABC, abstractmethod
+from abc import abstractmethod
 from collections.abc import Awaitable, Callable
 from functools import cached_property
-from typing import Any, Generic, TypeVar
+from typing import Any, Unpack
 
 from pydantic import BaseModel
+from typing_extensions import TypeVar
 
 from beeai_framework.agents.errors import AgentError
-from beeai_framework.agents.types import AgentMeta
+from beeai_framework.agents.types import AgentExecutionConfig, AgentMeta
+from beeai_framework.backend import AnyMessage
 from beeai_framework.context import Run, RunContext, RunMiddlewareType
 from beeai_framework.emitter import Emitter
 from beeai_framework.memory import BaseMemory
+from beeai_framework.runnable import Runnable, RunnableOptions, RunnableOutput
 from beeai_framework.utils import AbortSignal
 
-TOutput = TypeVar("TOutput", bound=BaseModel)
 
+class AgentOptions(RunnableOptions, total=False):
+    """Agent options."""
 
-class BaseAgent(ABC, Generic[TOutput]):
-    def __init__(self) -> None:
-        super().__init__()
+    expected_output: str | type[BaseModel]
+    """
+    Instruction for steering the agent towards an expected output format.
+    This can be a Pydantic model for structured output decoding and validation.
+    """
+
+    total_max_retries: int
+    """
+    Maximum number of model retries.
+    """
+
+    max_retries_per_step: int
+    """
+    Maximum number of model retries per step.
+    """
+
+    max_iterations: int
+    """
+    Maximum number of iterations.
```

Comment:
Either explain what iteration is to the user or remove the comment, as it adds no value.